site/content/3.12/aql/high-level-operations/upsert.md

```diff
@@ -63,6 +66,50 @@ UPDATE { logins: OLD.logins + 1 } IN users
 Note that in the `UPDATE` case it is possible to refer to the previous version of the
 document using the `OLD` pseudo-value.
 
+## Filter matching syntax
+
+The alternative syntax for `UPSERT` operations allows you to use more flexible
+filter conditions to look up documents. This means that the expression you use
+to look up a document does not have to be an object literal.
+
+<pre><code>UPSERT FILTER<em>filter-condition</em>
+INSERT <em>insertExpression</em>
+UPDATE <em>updateExpression</em>
+IN <em>collection</em></code></pre>
+
+<pre><code>UPSERT FILTER<em>filter-condition</em>
+INSERT <em>insertExpression</em>
+REPLACE <em>updateExpression</em>
+IN <em>collection</em></code></pre>
+
+The filter condition for the lookup can make use of the pseudo-variable
+`CURRENT` to access the lookup document and apply more filters on it than just
+equality matches. Example:
+
+```aql
+UPSERT FILTER CURRENT.name == 'superuser'
+INSERT { name: 'superuser', logins: 1, dateCreated: DATE_NOW() }
+UPDATE { logins: OLD.logins + 1 } IN users
+```
+
+While there can only be one `FILTER` statement in this version of `UPSERT`, the
+`FILTER` statement can also use operators such as `&&` and `||` to make
+more complex filter conditions.
+
+```aql
+UPSERT FILTER CURRENT.name == 'superuser' && CURRENT.active == true && (LOWER(CURRENT.member) == true || CURRENT.age == 33)
```

Comment:
Except for the LOWER(), you could use the exact matching syntax instead. Would be nicer to show a few different comparison operators. And say something about index utilization - I assume that the same limitations apply as with regular FILTER expressions (>, >=, ==, <, <=, IN, and IN_RANGE() may utilize indexes, but e.g. LIKE, !=, and ALL/ANY/NONE cannot)

Reply:
@jsteemann Do all FILTER expressions work here, but only can get accelerated by indexes?
Can you force an index hint for UPSERT to make it use an inverted index?

Reply:
All FILTER expressions should work here. An index can be used, but this is not a requirement.
All the optimizer will do for the lookup condition, is to generate a subquery as follows:
```
LET sub = (
  FOR temp IN collection 
    FILTER <filter condition as specified in upsert>
    LIMIT 1
    RETURN temp
)
```
The filter condition here is either the filter condition using the new syntax, or it is a filter condition that is auto-generated from the object literal (old syntax).

That subquery will work with or without indexes. If an appropriate index is found, the optimizer will use it. Otherwise it will do a full collection scan.

To the best of my knowledge (and according to some code I looked at), UPSERT supports index hints, but I haven't tried it myself yet. Let me know if I should try it out.

Reply:
Passing an index hint via UPSERT ... OPTIONS does show up in the execution plan for the generated subquery but it doesn't appear to be usable for inverted indexes - in maintainer builds, there is an assertion failure because "read own writes" is expected to be false by some inverted index code but it is actually set to true (as this is required for UPSERT). An error should be thrown in such a case: https://arangodb.atlassian.net/browse/BTS-1698